runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs

```diff
@@ -0,0 +1,363 @@
+    },
+};
+
+/// Stores state for rebuilding snapshot storages
+#[derive(Debug)]
+pub struct SnapshotStorageRebuilder {
+    /// Receiver for unpacked snapshot storage files
+    file_receiver: Receiver<PathBuf>,
+    /// Number of threads to rebuild with
+    num_threads: usize,
+    /// Snapshot storage lengths - from the snapshot file
+    snapshot_storage_lengths: HashMap<Slot, HashMap<usize, usize>>,
+    /// Container for storing snapshot file paths
+    storage_paths: DashMap<Slot, Mutex<Vec<PathBuf>>>,
+    /// Container for storing rebuilt snapshot storages
+    storage: DashMap<Slot, SlotStores>,
+    /// Tracks next append_vec_id
+    next_append_vec_id: Arc<AtomicAppendVecId>,
+    /// Tracker for number of processed slots
+    processed_slot_count: AtomicUsize,
+    /// Tracks the number of collisions in AppendVecId
+    num_collisions: AtomicUsize,
+}
+
+impl SnapshotStorageRebuilder {
+    /// Synchronously spawns threads to rebuild snapshot storages
+    pub fn rebuild_storage(
+        file_receiver: Receiver<PathBuf>,
+        num_threads: usize,
+        next_append_vec_id: Arc<AtomicAppendVecId>,
+    ) -> DashMap<Slot, SlotStores> {
+        let (snapshot_file_path, append_vec_files) = Self::get_snapshot_file(&file_receiver);
+        let snapshot_storage_lengths = Self::process_snapshot_file(snapshot_file_path).unwrap();
+        Self::spawn_rebuilder_threads(
+            file_receiver,
+            num_threads,
+            next_append_vec_id,
+            snapshot_storage_lengths,
+            append_vec_files,
+        )
+    }
+
+    /// Create the SnapshotStorageRebuilder for storing state during rebuilding
+    ///     - pre-allocates data for storage paths
+    fn new(
+        file_receiver: Receiver<PathBuf>,
+        num_threads: usize,
+        next_append_vec_id: Arc<AtomicAppendVecId>,
+        snapshot_storage_lengths: HashMap<Slot, HashMap<usize, usize>>,
+    ) -> Self {
+        let storage = DashMap::with_capacity(snapshot_storage_lengths.len());
+        let storage_paths: DashMap<_, _> = snapshot_storage_lengths
+            .iter()
+            .map(|(slot, storage_lengths)| {
+                (*slot, Mutex::new(Vec::with_capacity(storage_lengths.len())))
+            })
+            .collect();
+        Self {
+            file_receiver,
+            num_threads,
+            snapshot_storage_lengths,
+            storage_paths,
+            storage,
+            next_append_vec_id,
+            processed_slot_count: AtomicUsize::new(0),
+            num_collisions: AtomicUsize::new(0),
+        }
+    }
+
+    /// Waits for snapshot file
+    /// Due to parallel unpacking, we may receive some append_vec files before the snapshot file
+    /// This function will push append_vec files into a buffer until we receive the snapshot file
+    fn get_snapshot_file(file_receiver: &Receiver<PathBuf>) -> (PathBuf, Vec<PathBuf>) {
+        let mut append_vec_files = Vec::with_capacity(1024);
```

Comment:
what is 1024 here?

Reply:
pre-allocating the append_vec_files to some size to avoid resizing while we wait for the snapshot file.
this is probably larger than it needs to be though.

Reply:
add a comment for why '1024'